### ISSUE TITLE
Scss changes reloading, not injecting - proxy setup

### ISSUE THREAD

**Issue by igregson** [BrowserSync/gulp-browser-sync#16]
Browser Sync is doing a full page reload on ever save, including style files (`scss`).

I'm proxying to an Express server with the following gulp tasks:

**Browser Sync Task**

```
gulp.task('browser-sync', ['sass', 'hbs'], function() {
    return browserSync.init('./dev_themeRoot/**/*.*', {
      proxy: '127.0.0.1:2368',
      notify: false,
      ghostMode: {
          clicks: true,
          location: true,
          forms: true,
          scroll: false
      }
    });
});
```

**Sass task**

```
gulp.task('sass', function() {
  return gulp.src('./dev_themeRoot/styles/**/*.scss')
    .pipe($.plumber({
      errorHandler: onError
      }))
    .pipe($.sass({
      sourceComments: 'map',
      errLogToConsole: true,
      includePaths: [dir.stylesSrc + '/**/*.scss']
      }))
    .pipe($.flatten())
    .pipe(gulp.dest( themeRoot + '/assets/css' ))
    .pipe($.minifyCss())
    .pipe($.rename({suffix: '.min'}))
    .pipe($.minifyCss())
    .pipe(gulp.dest( themeRoot + '/assets/css' ))
    .pipe(browserSync.reload({stream:true}));
  });
```

**Watch task**

```
gulp.task('watch', ['sass', 'hbs', 'js', 'vendor', 'browser-sync'], function() {
  gulp.watch('./dev_themeRoot/styles/**/*.scss', ['sass']);
  gulp.watch('./dev_themeRoot/**/*.hbs', ['hbs']);
})
```

Ideas?

**Comment by daniele-orlando** [comment]
@justnorris Same identical issue like you.

**Comment by blakelapierre** [comment]
Getting the same results here. Injecting works perfectly when I'm just modifying LESS/CSS, but after a full reload happens (HTML or JS change), every style change forces a full page reload.

**Comment by shakyShane** [comment]
This is the recommended approach now for doing file watching + reload({stream: true})

https://github.com/BrowserSync/recipes/tree/master/recipes/gulp.sass

**Comment by shakyShane** [comment]
```js
// Static Server + watching scss/html files
gulp.task('serve', ['sass'], function() {

    browserSync({
        server: "./app"
    });

    gulp.watch(src.scss, ['sass']);
    gulp.watch(src.html).on('change', reload);
});

// Compile sass into CSS
gulp.task('sass', function() {
    return gulp.src(src.scss)
        .pipe(sass())
        .pipe(gulp.dest(src.css))
        .pipe(reload({stream: true}));
});
```

**Comment by blakelapierre** [comment]
That is pretty much identical to my setup, except I'm using LESS. I've verified that only one file (final concat'd css) is being piped to reload({stream: true}).

Like I said, the injection works fine until I force a full reload (with an HTML or JS change), then everytime after, a CSS change will also cause a full reload (although I briefly see the CSS injection happen successfully before the reload).

Right now I'm guessing it's an issue with reload() holding on to some kind of state from previous reloads. Will investigate.

**Comment by shakyShane** [comment]
There's no stat involved with the reload method, any issue is with the way gulp accepts functions in the task list.

that's why you should just use this pattern and it'll work fine.

```js
gulp.watch("*.html").on('change', browserSync.reload);
```

**Comment by blakelapierre** [comment]
Ok, found it.

Here was my watch section:

````
gulp.watch(src.html, [reload]);
  gulp.watch(src.scripts, ['js:debug', reload]);
  gulp.watch(src.templates, ['js:debug', reload]);
  gulp.watch(src.less, ['less:debug'])
      .on('change', event => {
        if (event.type === 'deleted') {
          delete cached.caches['less'][event.path];
          remember.forget('less', event.path);
        }
      });
````

For some reason that causes things to get "stuck" for subsequent runs that have nothing to do with those particular watches.

This was recommended....somewhere.

I changed to:

````
gulp.watch(src.html).on('change', reload);
  gulp.watch(src.scripts, ['js:debug']).on('change', reload);
  gulp.watch(src.templates, ['js:debug']).on('change', reload);
  gulp.watch(src.less, ['less:debug'])
      .on('change', event => {
        if (event.type === 'deleted') {
          delete cached.caches['less'][event.path];
          remember.forget('less', event.path);
        }
      });
````

And it's working fine.

It would be nice if the first way could be supported.

**Comment by blakelapierre** [comment]
Ok, one issue with this setup:

`gulp.watch(src.templates, ['js:debug']).on('change', reload);`

is that 'change' fires when the watch fires, so before/simultaneously with my task list. That means I miss out on whatever my task list produces.

I suppose the fix is the call reload({stream:true}) in your task's pipeline, but it would be nice to separate out reloading from the actual work of the task, which is what my previous call (mostly; until it stops working) did:

`gulp.watch(src.templates, ['js:debug', reload]);`

I'm not setup for debugging BrowserSync at the moment, but it seems that getting a stack trace from when https://github.com/BrowserSync/browser-sync/blob/master/lib/public/reload.js#L21 is called would help to figure out (and fix) what is going wrong.

**Comment by fredricgeorgsson-mobenga** [comment]
I have this problem as well. The style task streams styles via browserSync until a full build is done and a browsersync full reload is done. After that the style task reloads browsers instead of streaming them.

I've tried various ways to do the reload({ stream: true }) but the result is always the same, streams until a full reload has been made, just reloads after that.

The reciepe linked to @shakyShane does not really work since we do thing slike blakeapierre and have both watch and build, only watch should use browserSync.

gulpfile.js, trimmed a bit for readability.

```javascript
gulp.task('libs', function () {
    var folders = [
        'bower_components/jquery/dist/jquery.js',
        'bower_components/knockout/dist/knockout.debug.js',
        '!**/*.min.js'
    ];

    return gulp.src(folders)
        .pipe(size({
            title: 'libs',
            showFiles: true
        }))
        .pipe(sourcemaps.init())
        .pipe(concat('libs.js'))
        .pipe(gulpIf(node_env !== 'development', uglify()))
        .pipe(sourcemaps.write('../maps'))
        .pipe(gulp.dest(path.join(destDir, 'lib')));
});

gulp.task('styles', function () {
    var result = gulp.src(['./less/main.less', './features/**/*.less'])
        .pipe(size({
            title: 'less',
            'showFiles': true
        }))
        //.pipe(sourcemaps.init())
        .pipe(less({ compress: true }))
            .on('error', swallowError)
        //.pipe(sourcemaps.write('../maps'))
        .pipe(please({ 'autoprefixer': { 'browsers': ['> 1%', 'Android 2.3', 'iOS 5']} }))
        .pipe(concat('main.css'))
        .pipe(gulp.dest(path.join(destDir, 'css')));

    if (useBrowserSync) {
        return result.pipe(browserSync.reload({stream: true}))
    }

    return result;
});

gulp.task('template', function () {
    var opts = {
        cwd: destDir
    };

    var data = {
        libs: { files: ['lib/**/*.js'], opts: opts },
        css: { files: ['css/*.css'], opts: opts },
        script: { files: ['targets/' + node_env + '/ljhsdfsdf.js'], asContent: true }
    };

    return handlebarsStream('web/index.template.html', data)
        .pipe(rename('index.html'))
        .pipe(gulp.dest(destDir));
});

gulp.task('copy', function () {
    return gulp.src([
        'web/**/*',
        '!web/lib/**/*',
        '!web/css/**/*',
        '!web/index.template.html'
    ]).pipe(gulp.dest(destDir));
});

gulp.task('clean', function (done) {
    del([destDir], done);
});

gulp.task('browser-sync', function() {
    var port = '8080';

    if (argv.hasOwnProperty('node-app-port')) {
        port = argv['node-app-port'];
    }

    browserSync({
        proxy: 'localhost:' + port
    });
});

gulp.task('build', function (done) {
    runSequence(
        'clean',
        'jshint',
        'libs',
        'styles',
        'copy',
        'template',
        done);
});

gulp.task('watch', ['build', 'browser-sync'], function () {
    useBrowserSync = true;
    gulp.watch('web/**', ['build', browserSync.reload]);
    gulp.watch('mock/**', ['build', browserSync.reload]);
    gulp.watch('features/**', ['build', browserSync.reload]);
    gulp.watch('less/**', ['styles']);
    gulp.watch('targets/' + node_env + '/**', ['build', browserSync.reload]);
});

gulp.task('default', ['build']);
```

**Comment by flip4bytes** [comment]
Does BrowserSync inject work with Stylus? Because everything I try, even when I duplicate the spec, it doesn't inject the changes. This is what I have:

```
var gulp = require('gulp'),
    browserSync = require('browser-sync'),
    reload = browserSync.reload;    
    prefix = require('gulp-autoprefixer'),
    filter = require('gulp-filter'),
    plumber = require('gulp-plumber'),
    stylus = require('gulp-stylus'),


gulp.task('browser-sync', ['styles'], function() {
  browserSync({
    files: "themes/csbzone/assets/css/*.css",
    proxy: "localhost.csbzone.com",
    notify: false,
    open: false
  });
});

gulp.task('styles', function(){
  return gulp.src('themes/site/assets/styl/theme.styl') // Main theme.styl file
    .pipe(plumber({
      errorHandler: onError
    }))
    .pipe(stylus({
      compress: true, // Compresses the CSS
    }))
    .pipe(gulp.dest('themes/site/assets/css')) // Write to css dir
    .pipe(filter('**/*.css')) // Filter the stream to ensure only CSS files passed.
    .pipe(reload({stream: true})); // Inject into browsers
});

// gulp.watch('themes/site/assets/styl/**/*.styl', ['styles']);

gulp.task('default', ['styles', 'scripts', 'browser-sync'], function() {
  gulp.watch('themes/site/assets/styl/**/*.styl', ['styles']);
  gulp.watch('themes/site/**/*.htm').on('change', reload);
});
```

**Comment by craigmdennis** [comment]
I'm also experiencing this issue. I'm using Jade so have to wait for all files to compile before triggering `reload`. If I use the above method, BrowserSync triggers a reload for each file that is compiled.

I had to use https://github.com/juanfran/gulp-jade-inheritance which seems a little verbose but has done the trick.

**Comment by shakyShane** [comment]
@craigmdennis - study the [gulpfile.js](https://github.com/BrowserSync/recipes/blob/master/recipes/gulp.jade/gulpfile.js) in the official [Gulp, Sass + Jade Recipe](https://github.com/BrowserSync/recipes/tree/master/recipes/gulp.jade) - it shows how to correctly wait for jade compilation to complete before reloading.

tl;dr, for everyone else - create a separate task for calling your jade task & always return your streams.

```js
/**
 * Compile jade files into HTML
 */
gulp.task('templates', function() {

    var YOUR_LOCALS = {};

    return gulp.src('./app/*.jade')
        .pipe(jade({
            locals: YOUR_LOCALS
        }))
        .pipe(gulp.dest('./dist/'))
});

/**
 * Important!!
 * Separate task for the reaction to `.jade` files
 */
gulp.task('jade-watch', ['templates'], reload);

/**
 * Serve and watch the scss/jade files for changes
 */
gulp.task('default', ['sass', 'templates'], function () {

    browserSync({server: './dist'});
    gulp.watch('./app/*.jade',      ['jade-watch']);
});
```

**Comment by dangelion** [comment]
Hi
my gulpfile works right except for this: when I modify a SASS file it reload the page instead of injecting CSS.

Here my gulpfile, what's wrong?

```
// /**
//  * Gulp
//  * -----------------------------------------------------------------------------
//  */
'use strict';

/* Plugins
 * -------------------- */
var gulp = require('gulp');
var autoprefixer = require('gulp-autoprefixer');
var sass = require('gulp-sass');
var sourcemaps = require('gulp-sourcemaps');
var uglify = require('gulp-uglify');
var browserify = require('browserify');
var source = require('vinyl-source-stream');
var argv = require('yargs').argv;
var gulpif = require('gulp-if');
var buffer = require('vinyl-buffer');
var imagemin = require('gulp-imagemin');
var pngquant = require('imagemin-pngquant');
var changed = require('gulp-changed');
var browserSync = require('browser-sync');
var concat = require('gulp-concat');


/*
 * Settings
 */
var projectUrl = "http://myproject.dev:9999";



/*
 * Tasks
 */

/* Task for building the Sass files
 * -------------------- */
gulp.task('styles', function() {
    var format = argv.production ? 'compressed' : 'expanded';

    return gulp.src('source/sass/*.scss')
        .pipe(gulpif(!argv.production, sourcemaps.init()))
        .pipe(sass({
            outputStyle: format
        }))
        .on('error', function(err) {
            console.error(err.message);
        })
        .pipe(autoprefixer({
            browsers: ['last 8 versions', 'ie 8', 'ie 9']
        }))
        .pipe(gulpif(!argv.production, sourcemaps.write('.')))
        .pipe(gulp.dest('dist/css/'))
        .pipe(browserSync.reload({stream:true}))
});


/* Scripts
 * -------------------- */
gulp.task('scripts', function() {
    return gulp.src([
            //'source/scripts/**/*.js',
            'bower_components/jquery/dist/jquery.js',
            'bower_components/vide/dist/jquery.vide.js',
            'bower_components/Morphist/dist/morphist.js',
            'bower_components/velocity/velocity.js',
            'bower_components/velocity/velocity.ui.js',
            'bower_components/fullpage.js/dist/jquery.fullpage.js',
            'bower_components/fullpage.js/vendors/jquery.slimscroll.js',
            'source/js/components/common.js',
            'source/js/pages/*.js',
            'source/js/main.js',
           ])
        .pipe(concat('script.js'))
        .pipe(buffer())
        .pipe(gulpif(!argv.production, sourcemaps.init({loadMaps: true})))
        .pipe(gulpif(!argv.production, sourcemaps.write('./')))
        .pipe(gulpif(argv.production, uglify()))
        .pipe(gulp.dest('dist/js/'))
        .pipe(browserSync.reload({stream:true}))
});


/* Image compression
 * -------------------- */
gulp.task('images', function () {
    return gulp.src('source/img/**/*')
        .pipe(changed('dist/img'))
        .pipe(imagemin({
            progressive: true,
            svgoPlugins: [{removeViewBox: false}],
            use: [pngquant()]
        }))
        .pipe(gulp.dest('dist/img'));
});


/* Video
 * -------------------- */
gulp.task('video', function () {
    return gulp.src('source/video/**/*')
        .pipe(changed('dist/video'))
        .pipe(gulp.dest('dist/video'));
});


/* Fonts
 * -------------------- */
gulp.task('fonts', function () {
    return gulp.src('source/fonts/**/*')
        .pipe(changed('dist/fonts'))
        .pipe(gulp.dest('dist/fonts'));
});


/* BrowserSync
 * Configs from http://quenchjs.com/
 * -------------------- */
gulp.task('browser-sync', function() {
  browserSync({
    proxy: {
      // Vhost-based url
      target: projectUrl,
      // Enables websocket
      ws: true
    },
  });
});

gulp.task('bs-reload', function () {
  browserSync.reload();
});


/*
 
 * -------------------- */
gulp.task('default', ['build', 'watch', 'browser-sync']);
gulp.task('build', ['scripts', 'styles', 'images', 'fonts', 'video']);
gulp.task('watch', function() {
    gulp.watch(['**/*.html', '**/*.php'], ['bs-reload']);
    gulp.watch('source/js/**/*.js', ['scripts']);
    gulp.watch('source/sass/**/*.scss', ['styles']);
    gulp.watch('source/img/*', ['images']);
    gulp.watch('source/video/*', ['video']);
    gulp.watch('source/fonts/*', ['fonts']);
});
```
